README.md
[![FIWARE Big BangBanner](https://raw.githubusercontent.com/lets-fiware/FIWARE-Big-Bang/gh-pages/images/FIWARE-Big-Bang-non-free.png)](https://www.letsfiware.jp/)
[![NGSI v2](https://img.shields.io/badge/NGSI-v2-5dc0cf.svg)](https://fiware-ges.github.io/orion/api/v2/stable/)

![FIWARE: Tools](https://nexus.lab.fiware.org/repository/raw/public/badges/chapters/deployment-tools.svg)
[![License: MIT](https://img.shields.io/github/license/lets-fiware/FIWARE-Big-Bang.svg)](https://opensource.org/licenses/MIT)
[![Support badge](https://img.shields.io/badge/tag-fiware-orange.svg?logo=stackoverflow)](https://stackoverflow.com/questions/tagged/fiware+fi-bb)
<br/>
[![Lint](https://github.com/lets-fiware/FIWARE-Big-Bang/actions/workflows/lint.yml/badge.svg)](https://github.com/lets-fiware/FIWARE-Big-Bang/actions/workflows/lint.yml)
[![codecov](https://codecov.io/gh/lets-fiware/FIWARE-Big-Bang/branch/main/graph/badge.svg?token=OHFTT6TUIS)](https://codecov.io/gh/lets-fiware/FIWARE-Big-Bang)
<br/>
[![Ubuntu 18.04](https://github.com/lets-fiware/FIWARE-Big-Bang/actions/workflows/ubuntu-18.04.yml/badge.svg)](https://github.com/lets-fiware/FIWARE-Big-Bang/actions/workflows/ubuntu-18.04.yml)
[![Ubuntu 20.04](https://github.com/lets-fiware/FIWARE-Big-Bang/actions/workflows/ubuntu-20.04.yml/badge.svg)](https://github.com/lets-fiware/FIWARE-Big-Bang/actions/workflows/ubuntu-20.04.yml)
<br/>

The FIWARE Big Bang is a turnkey solution for setting up a FIWARE instance in the cloud.

| :books: [Documentation](https://fi-bb.letsfiware.jp/) | :dart: [Roadmap](./ROADMAP.md) |
|-------------------------------------------------------|--------------------------------|

## What is FIWARE Big Bang?

> I am at all events convinced that He does not play dice.
>
> — Albert Einstein

The FIWARE Big Bang allows you to install FIWARE Generic enablers easily into your virtual machine in the cloud.
FI-BB stands for FIWARE Big Bang.

## Requirements

-   A virtual machine with a global IP address
-   A domain name
-   Supported Linux distribution:
    -   Ubuntu 20.04 (Recommended Linux distribution)
    -   Ubuntu 18.04
    -   CentOS 7, 8
-   Supported FIWARE GEs:
    -   Keyrock
    -   Wilma
    -   Orion
    -   Cygnus
    -   Comet
    -   QuantumLeap
    -   WireCloud
    -   Ngsiproxy
    -   IoT Agent for UltraLight
-   Supported third party OSS:
    -   Node-RED
    -   Grafana
    -   Mosquitto
    -   Knowage

## Prerequisite

Before running the setup script, you need to register sub-domain names of FIWARE GEs that you want to use as shown:

-   keyrock.letsfiware.jp
-   orion.letsfiware.jp
-   wirecloud.letsfiware.jp
-   ngsiproxy.letsfiware.jp

## Getting Started

Download a tar.gz file for the FIWARE Big Bang and run the `lets-fiware.sh` script with your domain name.

```
curl -sL https://github.com/lets-fiware/FIWARE-Big-Bang/archive/refs/tags/v0.6.0.tar.gz | tar zxf -
```

```
cd FIWARE-Big-Bang-0.6.0/
./lets-fiware.sh letsfiware.jp
```

## Why is it named FIWARE Big Bang?

The name of this product comes from the Big Bang theory of the universe. Because most FIWARE generic enablers in
the Core Context Management chapter have an astrology name and this product creates a FIWARE instance as your own
universe in which various FIWARE GEs run.

## Copyright and License

Copyright (c) 2021 Kazuhito Suda<br>
Licensed under the [MIT License](./LICENSE).


CHANGELOG.md
## FIWARE Big Bang v0.6.0-next

-   Integrate Knowage (#XX)
-   Manage installation state (#84)
-   Add pre-install widgets (#83)

## FIWARE Big Bang v0.6.0 - 26 Oct, 2021

-   Update Orion to version 3.3.0 (#80)
-   Add postfix (#79)
-   Add Regproxy (#78)
-   Improve creating docker images (#77)
-   Add Queryproxy (#76)

## FIWARE Big Bang v0.5.0 - 9 Oct, 2021

-   Hardening Node-RED support (#73)
-   Improve coverage (#72)
-   Replace httpRoot with httpAdminRoot and httpNodeRoot for Node-RED (#71)
-   Improve coverage (#70)
-   Support Node-RED multi instance (#69)

## FIWARE Big Bang v0.4.0 - 3 Oct, 2021

-   Hardening IoT Agent for UltraLight over TLS (#66)
-   Fix no coverage report found (#65)
-   Improve WireCloud installation test (#64)
-   Improve coverage (#63)
-   Add IoT Agent for UltraLight on MQTT (#62)
-   Fix WireCloud setup (#61)
-   Add environment variables for logging level (#60)
-   Hardening WireCloud installation (#59)
-   Add persistent volume for Mongodb (#58)

## FIWARE Big Bang v0.3.0 - 25 Sep, 2021

-   Fix certbot image (#55)
-   Fix WireCloud settings for NGSI Go (#54)
-   Improve documentation (#53)
-   Add Keyrock with PostgreSQL (#52)
-   Add default server for nginx (#51)
-   Improve coverage (#50)
-   Add certmock (#49)
-   Update README.md (#48)
-   Refactor (#47)

## FIWARE Big Bang v0.2.0 - 20 Sep, 2021

-   Improve docker install on Ubuntu (#44)
-   Add NGSI Go bash autocomplete (#43)
-   Fix cert test case (#42)
-   Fix command parameter for certbot renew and revoke (#41)
-   Add feature to install WireCloud widgets (#40)
-   Add make command (#39)
-   Fix certbot option (#38)
-   Improve coverage (#37)
-   Add coverage (#36)
-   Refactor (#35)
-   Add yamllint feature (#34)
-   Add shellcheck feature (#33)
-   Fix typos in README.md (#32)
-   Add status badges (#31)
-   Improve linting process (#30)
-   Add FIWARE Big Bang logo mark (#29)
-   Improve admin user definition (#28)
-   Add role-based access control to Node-RED (#27)
-   Replace nginx log directory with /var/log/fiware/nginx (#26)
-   Fix server certificate files for setup test (#25)
-   Improve cert process (#24)
-   Add bind-utils for CentOS (#23)
-   Add GitHub Actions Badge (#22)
-   Add setup test to GitHub Actions (#21)
-   Update Orion version to 3.2.0 (#20)
-   Fix owner of log directory on CentOS (#19)
-   Update config.sh (#18)
-   Fix owner, group and mode bits of log files on CentOS (#17)
-   Add syslog settings (#16)
-   Add firewall settings (#15)
-   Fix installing Docker on Ubuntu (#14)
-   Improve Nginx configuration (#13)
-   Add NGSI Go settings (#12)
-   Fix template for nginx (#11)
-   Improve Configuration (#10)
-   Integrate QuantumLeap (#9)
-   Fix missing parameters for certbot (#8)
-   Fix missing pwgen command (#7)
-   Add Admin role for WireCloud (#4)
-   Update template for nginx (#3)
-   Update default settings (#1)

## FIWARE Big Bang v0.1.0 - 2 Sep, 2021

-   Initial release


config.sh
#!/bin/bash

#
# Required parameters
#
KEYROCK=keyrock
ORION=orion

#
# Optional parameters
#

# Set a sub-domain name of the FIWARE GE you want to set.
COMET=
QUANTUMLEAP=
WIRECLOUD=
NGSIPROXY=
IOTAGENT=

NODE_RED=
GRAFANA=
MOSQUITTO=
KNOWAGE=knowage

# Use queryproxy. (true or false) Default: false
QUERYPROXY=

# Use MQTT (Port 1883). (true or false) Default: false
MQTT_1883=

# Use MQTT TLS (Port 8883). (true or false) Default: true
MQTT_TLS=

# Use PostgreSQL as back-end database for Keyrock. (true or false) Default: false
KEYROCK_POSTGRES=

# Set username for MQTT. Defaut: fiware
MQTT_USERNAME=

# Set password for MQTT. Defaut: automatically generated
MQTT_PASSWORD=

# Set a name of an admin user for Keyrock. Default: admin
IDM_ADMIN_USER=

# Set an e-mail address of an admin user for Keyrock. Default: IDM_ADMIN_NAME @ DOMAIN_NAME
IDM_ADMIN_EMAIL=

# Set a password of an admin user for Keyrock. Default: automatically generated
IDM_ADMIN_PASS=

# Enable firewall. (true or false) Default: false
FIREWALL=

# Node-RED multi instance
# Number of Node-RED instance. default: 1
NODE_RED_INSTANCE_NUMBER=

# username for Node-RED instance. default: node-red
NODE_RED_INSTANCE_USERNAME=

# httpNodeRoot for Node-RED instance. default: / or /node-red???
NODE_RED_INSTANCE_HTTP_NODE_ROOT=

# httpAdminRoot for Node-RED instance. default: / or /node-red???
NODE_RED_INSTANCE_HTTP_ADMIN_ROOT=

# Use Postfix (local delivery). (true or false) Default: false
POSTFIX=

#
# Regproxy options
#
# Use regproxy. (true or false) Default: false
REGPROXY=

# NgsiType for remote broker. (v2 or ld) Default: v2
REGPROXY_NGSITYPE=

# Host for remote broker.
REGPROXY_HOST=

# IdM type for remote broker.
REGPROXY_IDMTYPE=

# IdM host for remote broker.
REGPROXY_IDMHOST=

# username for remote broker.
REGPROXY_USERNAME=

# password for remote broker.
REGPROXY_PASSWORD=

# client id for remote broker.
REGPROXY_CLIENT_ID=

# client secret for remote broker.
REGPROXY_CLIENT_SECRET=

#
# Certbot options
#
# Set a e-mail address for certbot. Defaul: a e-mail address of an admin user for Keyrock.
CERT_EMAIL=

# Revoke and reacquire the certificate. (true or false) Default: false
CERT_REVOKE=

# Use --test-cert option.
# CERT_TEST=--test-cert
CERT_TEST=

# Use --force-renewal option. (true or false) Default: false
CERT_FORCE_RENEWAL=

# Docker images
IMAGE_KEYROCK=fiware/idm:8.1.0
IMAGE_WILMA=fiware/pep-proxy:8.1.0
IMAGE_ORION=telefonicaiot/fiware-orion:3.3.0
IMAGE_CYGNUS=fiware/cygnus-ngsi:2.10.0
IMAGE_COMET=fiware/sth-comet:2.8.0
IMAGE_WIRECLOUD=fiware/wirecloud:1.3.1
IMAGE_NGSIPROXY=fiware/ngsiproxy:1.2.2
IMAGE_QUANTUMLEAP=orchestracities/quantumleap:0.8.1
IMAGE_IOTAGENT=fiware/iotagent-ul:1.16.2

IMAGE_MONGO=mongo:4.4
IMAGE_MYSQL=mysql:5.7
IMAGE_POSTGRES=postgres:11
IMAGE_CRATE=crate:4.1.4

IMAGE_TOKENPROXY=letsfiware/tokenproxy:0.6.0-next
IMAGE_QUERYPROXY=letsfiware/queryproxy:0.6.0-next
IMAGE_REGPROXY=letsfiware/regproxy:0.6.0-next

IMAGE_NGINX=nginx:1.21
IMAGE_REDIS=redis:6
IMAGE_ELASTICSEARCH=elasticsearch:2.4
IMAGE_MEMCACHED=memcached:1
IMAGE_GRAFANA=grafana/grafana:6.1.6
IMAGE_MOSQUITTO=eclipse-mosquitto:1.6
IMAGE_NODE_RED=letsfiware/node-red:0.6.0-next
IMAGE_POSTFIX=letsfiware/postfix:0.6.0-next

IMAGE_CERTBOT=certbot/certbot:v1.18.0

# Logging option

# Keyrock
#   https://github.com/ging/fiware-idm/blob/master/doc/installation_and_administration_guide/environment_variables.md
#   true and false
IDM_DEBUG=false

# Cygnus
#   https://github.com/telefonicaid/fiware-cygnus/blob/master/doc/cygnus-common/installation_and_administration_guide/install_with_docker.md
#   INFO, DEBUG
CYGNUS_LOG_LEVEL=INFO

# Comet
#   https://github.com/telefonicaid/fiware-sth-comet/blob/master/doc/manuals/running.md
#   Possible values are: "DEBUG", "INFO", "WARN", "ERROR" and "FATAL" 
LOGOPS_LEVEL=INFO

# Quantumleap
#   INFO, DEBUG
QUANTUMLEAP_LOGLEVEL=INFO

# WireCloud
#   https://github.com/Wirecloud/docker-wirecloud/blob/master/README.md
WIRECLOUD_LOGLEVEL=INFO

# Tokenproxy
TOKENPROXY_LOGLEVEL=info
TOKENPROXY_VERBOSE=

# Queryproxy
QUERYPROXY_LOGLEVEL=info

# Regproxy log level
REGPROXY_LOGLEVEL=info

# Regproxy verbose
REGPROXY_VERBOSE=false

# Node-RED
#   https://nodered.org/docs/user-guide/runtime/logging
NODE_RED_LOGGING_LEVEL=info
NODE_RED_LOGGING_METRICS=
NODE_RED_LOGGING_AUDIT=

# Grafana
#   https://grafana.com/docs/grafana/latest/administration/configuration/#configure-with-environment-variables
#   https://github.com/grafana/grafana/blob/main/conf/defaults.ini
#   "debug", "info", "warn", "error", "critical"
GF_LOG_LEVEL=info

#   https://mosquitto.org/man/mosquitto-conf-5.html
#   debug, error, warning, notice, information, subscribe, unsubscribe, websockets, none, all
MOSQUITTO_LOG_TYPE=error,warning,notice,information


setup/knowage.sh
#!/bin/bash
set -e

# usage: file_env VAR [DEFAULT]
#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
file_env() {
	local var="$1"
	local fileVar="${var}_FILE"
	local def="${2:-}"
	if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
		echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
		exit 1
	fi
	local val="$def"
	if [ "${!var:-}" ]; then
		val="${!var}"
	elif [ "${!fileVar:-}" ]; then
		val="$(< "${!fileVar}")"
	fi
	export "$var"="$val"
	unset "$fileVar"
}

change_value() {
	touch $INIT_PROP_FILE_TEMP
	local label="$1"
	local value_check="$2"
	tr -d '\n' < $INIT_PROP_FILE | sed 's#/>#/>\n#g' | sed 's/valueCheck=/\x00/g' | sed -E "s#(label=\"${label}\"[^\x00]*\x00)\"[^\"]*#\1\"${value_check}#g" | sed 's/\x00/valueCheck=/g' > $INIT_PROP_FILE_TEMP && mv $INIT_PROP_FILE_TEMP $INIT_PROP_FILE
}

file_env "DB_HOST"
file_env "DB_PORT"
file_env "DB_DB"
file_env "DB_USER"
file_env "DB_PASS"

file_env "CACHE_DB_HOST"
file_env "CACHE_DB_PORT"
file_env "CACHE_DB_DB"
file_env "CACHE_DB_USER"
file_env "CACHE_DB_PASS"

file_env "AJP_SECRET"

file_env "HAZELCAST_HOSTS"
file_env "HAZELCAST_PORT"

# Wait for MySql
./wait-for-it.sh ${DB_HOST}:${DB_PORT} -- echo "MySql is up!"

# Placeholder created after the first boot of the container
CONTAINER_INITIALIZED_PLACEHOLDER=/.CONTAINER_INITIALIZED

# Check if this is the first boot
if [ ! -f "$CONTAINER_INITIALIZED_PLACEHOLDER" ]
then
	file_env "HMAC_KEY"
	file_env "PASSWORD_ENCRYPTION_SECRET"
	file_env "PUBLIC_ADDRESS"

        if [ "${OAUTH2}" = "keyrock" ]; then

                INIT_PROP_FILE=${KNOWAGE_DIRECTORY}/${APACHE_TOMCAT_PACKAGE}/webapps/knowage/WEB-INF/classes/it/eng/spagobi/commons/initializers/metadata/config/configs.xml
                INIT_PROP_FILE_TEMP=${KNOWAGE_DIRECTORY}/webapps/knowage/WEB-INF/classes/it/eng/spagobi/commons/initializers/metadata/config/configs.xml.temp
                SERVER_CONF=${KNOWAGE_DIRECTORY}/conf/server.xml
                WEB_XML=${KNOWAGE_DIRECTORY}/webapps/knowage/WEB-INF/web.xml
                KNOWAGE_JAR=${KNOWAGE_DIRECTORY}/webapps/knowage/WEB-INF/lib/knowage-utils-7.0.0.jar
                UNZIPPED_JAR=knowageJAR
                KNOWAGE_CONF=${UNZIPPED_JAR}/it/eng/spagobi/security/OAuth2/configs.properties
                INITIALIZER_XML=${KNOWAGE_DIRECTORY}/webapps/knowage/WEB-INF/conf/config/initializers.xml

	        change_value "SPAGOBI_SSO.ACTIVE" "true"
	        change_value "SPAGOBI.SECURITY.PORTAL-SECURITY-CLASS.className" "it.eng.spagobi.security.OAuth2SecurityInfoProvider"
	        change_value "SPAGOBI.SECURITY.USER-PROFILE-FACTORY-CLASS.className" "it.eng.spagobi.security.OAuth2SecurityServiceSupplier"
	        change_value "SPAGOBI_SSO.SECURITY_LOGOUT_URL" "https://keyrock.mydomain.de/auth/external_logout"
	
	        sed -i "s/it.eng.spagobi.services.common.FakeSsoService/it.eng.spagobi.services.oauth2.Oauth2SsoService/g" $SERVER_CONF
	        sed -i "s/it.eng.spagobi.commons.initializers.metadata.MetadataInitializer/it.eng.spagobi.commons.initializers.metadata.OAuth2MetadataInitializer/g" $INITIALIZER_XML
	        sed -i "s/<!-- START OAUTH 2/<!-- START OAUTH 2 -->/g" $WEB_XML
	        sed -i "s/END OAUTH 2 -->/<!-- END OAUTH 2 -->/g" $WEB_XML

	        unzip $KNOWAGE_JAR -d $UNZIPPED_JAR
	        sed -i "s/CLIENT_ID.*/CLIENT_ID=${CLIENT_ID}/g" $KNOWAGE_CONF
	        sed -i "s/SECRET.*/SECRET=${CLIENT_SECRET}/g" $KNOWAGE_CONF
	        sed -i "s#AUTHORIZE_URL.*#AUTHORIZE_URL=${KEYROCK_URL}/oauth2/authorize#g" $KNOWAGE_CONF
	        sed -i "s#ACCESS_TOKEN_URL.*#ACCESS_TOKEN_URL=${KEYROCK_URL}/oauth2/token#g" $KNOWAGE_CONF
	        sed -i "s#USER_INFO_URL.*#USER_INFO_URL=${KEYROCK_URL}/user#g" $KNOWAGE_CONF
	        sed -i "s#REDIRECT_URI.*#REDIRECT_URI=${KEYROCK_REDIRECT_URI}#g" $KNOWAGE_CONF
	        sed -i "s#TOKEN_PATH.*#TOKEN_PATH=${KEYROCK_TOKEN_PATH}#g" $KNOWAGE_CONF

                sed -i "s#REST_BASE_URL.*#REST_BASE_URL=${KEYROCK_URL}/#g" $KNOWAGE_CONF
	
	        sed -i "s#APPLICATION_ID.*#APPLICATION_ID=${KEYROCK_APPLICATION_ID}#g"
	        sed -i "s/ADMIN_ID.*/ADMIN_ID=${KEYROCK_ADMIN_ID}/g"
	        sed -i "s/ADMIN_EMAIL.*/ADMIN_EMAIL=${KEYROCK_ADMIN_EMAIL}/g"
	        sed -i "s/ADMIN_PASSWORD.*/ADMIN_PASSWORD=${KEYROCK_ADMIN_PASSWORD}/g"
	        cd $UNZIPPED_JAR; zip -r $KNOWAGE_JAR *
	        cd ..
	
        fi

	# Generate default values for the optional env vars
	if [[ -z "$PUBLIC_ADDRESS" ]]
	then
	        #get the address of container
	        #example : default via 172.17.42.1 dev eth0 172.17.0.0/16 dev eth0 proto kernel scope link src 172.17.0.109
	        PUBLIC_ADDRESS=`ip route | grep src | awk '{print $9}'`
	fi
	
	if [ -z "$HMAC_KEY" ]
	then
		echo "The HMAC_KEY environment variable is needed"
		exit -1
	fi
	
	if [ -z "$PASSWORD_ENCRYPTION_SECRET" ]
	then
		echo "The PASSWORD_ENCRYPTION_SECRET environment variable is needed"
		exit -1
	fi
	
	if [ -z "$DB_HOST" ] || [ -z "$DB_PORT" ] || [ -z "$DB_DB"   ] || [ -z "$DB_USER" ] || [ -z "$DB_PASS" ]
	then
		echo "The DB_HOST, DB_PORT, DB_DB, DB_USER, DB_PASS environment variables are needed"
		exit -1
	fi

	if [ -z "$CACHE_DB_HOST" ] || [ -z "$CACHE_DB_PORT" ] || [ -z "$CACHE_DB_DB"   ] || [ -z "$CACHE_DB_USER" ] || [ -z "$CACHE_DB_PASS" ]
	then
		echo "The CACHE_DB_HOST, CACHE_DB_PORT, CACHE_DB_DB, CACHE_DB_USER, CACHE_DB_PASS environment variables are needed"
		exit -1
	fi

	if [ -z "$AJP_SECRET" ]
	then
		AJP_SECRET=$( openssl rand -base64 32 )
		echo "###################################################################"
		echo "#"
		echo "# Random generated AJP secret:"
		echo "#   ${AJP_SECRET}"
		echo "#"
	fi

	if [ -z "$HAZELCAST_HOSTS" ]
	then
		HAZELCAST_HOSTS="127.0.0.1"
		echo "###################################################################"
		echo "#"
		echo "# The HAZELCAST_HOSTS environment not present. Knowage will launch "
		echo "# one internally."
		echo "#"
	fi

	if [ -z "$HAZELCAST_PORT" ]
	then
		HAZELCAST_PORT="5701"
	fi

	# Replace the address of container inside server.xml
	sed -i "s|http:\/\/.*:8080|http:\/\/${PUBLIC_ADDRESS}:8080|g" ${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	sed -i "s|http:\/\/.*:8080\/knowage|http:\/\/localhost:8080\/knowage|g" ${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	sed -i "s|http:\/\/localhost:8080|http:\/\/${PUBLIC_ADDRESS}:8080|g" ${KNOWAGE_DIRECTORY}/apache-tomcat/webapps/knowage/WEB-INF/web.xml
	
	# Insert knowage metadata into db if it doesn't exist
	result=`mysql -h${DB_HOST} -P${DB_PORT} -u${DB_USER} -p${DB_PASS} ${DB_DB} -e "SHOW TABLES LIKE '%SBI_%';"`
	if [ -z "$result" ]; then
		mysql -h${DB_HOST} -P${DB_PORT} -u${DB_USER} -p${DB_PASS} ${DB_DB} --execute="source ${MYSQL_SCRIPT_DIRECTORY}/MySQL_create.sql"
		mysql -h${DB_HOST} -P${DB_PORT} -u${DB_USER} -p${DB_PASS} ${DB_DB} --execute="source ${MYSQL_SCRIPT_DIRECTORY}/MySQL_create_quartz_schema.sql"
	fi
	
	# Set DB connection for Knowage metadata
	xmlstarlet ed -P -L \
		-u "//Server/GlobalNamingResources/Resource[@name='jdbc/knowage']/@url"      -v "jdbc:mysql://${DB_HOST}:${DB_PORT}/${DB_DB}" \
		-u "//Server/GlobalNamingResources/Resource[@name='jdbc/knowage']/@username" -v "${DB_USER}" \
		-u "//Server/GlobalNamingResources/Resource[@name='jdbc/knowage']/@password" -v "${DB_PASS}" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	
	# Set DB connection for Knowage cache
	xmlstarlet ed -P -L \
		-u "//Server/GlobalNamingResources/Resource[@name='jdbc/ds_cache']/@url"      -v "jdbc:mysql://${CACHE_DB_HOST}:${CACHE_DB_PORT}/${CACHE_DB_DB}" \
		-u "//Server/GlobalNamingResources/Resource[@name='jdbc/ds_cache']/@username" -v "${CACHE_DB_USER}" \
		-u "//Server/GlobalNamingResources/Resource[@name='jdbc/ds_cache']/@password" -v "${CACHE_DB_PASS}" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml

	# Set HMAC key
	xmlstarlet ed -P -L \
		-u "//Server/GlobalNamingResources/Environment[@name='hmacKey']/@value" -v "${HMAC_KEY}" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	
	# Set password encryption key
	echo $PASSWORD_ENCRYPTION_SECRET > ${KNOWAGE_DIRECTORY}/apache-tomcat/conf/passwordEncryptionSecret

	# Set AJP secret
	xmlstarlet ed -P -L \
		-d "//Server/Service/Connector[contains(@protocol, 'AJP')]/@secretRequired" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	xmlstarlet ed -P -L \
		-d "//Server/Service/Connector[contains(@protocol, 'AJP')]/@secret" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	xmlstarlet ed -P -L \
		-i "//Server/Service/Connector[contains(@protocol, 'AJP')]" -t attr -n secretRequired -v "true" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	xmlstarlet ed -P -L \
		-i "//Server/Service/Connector[contains(@protocol, 'AJP')]" -t attr -n secret -v "${AJP_SECRET}" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/server.xml
	
	# Setting hazelcast.xml
	# 
	# N.B.: the _ in Xmlstarlet XPath stands for default namespace

	# Set port
	xmlstarlet ed -P -L \
		-u "/_:hazelcast/_:network/_:port" -v ${HAZELCAST_PORT} \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/hazelcast.xml

	# Clean up the member list
	xmlstarlet ed -P -L \
		-d "/_:hazelcast/_:network/_:join/_:tcp-ip/_:member-list/_:member" \
		${KNOWAGE_DIRECTORY}/apache-tomcat/conf/hazelcast.xml
	
	# Set the actual member list
	echo -n "${HAZELCAST_HOSTS}" | xargs -d "," -n 1 -i"{}" \
		xmlstarlet ed -P -L \
			-s "/_:hazelcast/_:network/_:join/_:tcp-ip/_:member-list" -t elem -n member -v \{\} \
			${KNOWAGE_DIRECTORY}/apache-tomcat/conf/hazelcast.xml

	# Format
	xmlstarlet ed -O -L ${KNOWAGE_DIRECTORY}/apache-tomcat/conf/hazelcast.xml

	# Create the placeholder to prevent multiple initializations
	touch "$CONTAINER_INITIALIZED_PLACEHOLDER"
fi

exec "$@"
